packages/indicators/src/builders/buildArithmetic.ts

```diff
@@ -29,18 +37,53 @@ const validateConfig = (config: ArithmeticConfig) => {
   });
 };
 
+const analyticsToAnalyticalEvents = (analytics: Analytic[]): AnalyticalEvent[] => {
+  const eventMap: Record<string, AnalyticalEvent> = {};
+  analytics.forEach(analytic => {
+    const { dataElement, organisationUnit, period, value } = analytic;
+    const key = `${organisationUnit}__${period}`;
+    if (!eventMap[key]) {
+      eventMap[key] = { organisationUnit, period, dataValues: {} };
+    }
+    eventMap[key].dataValues[dataElement] = value;
+  });
+
+  return Object.values(eventMap);
+};
+
+const fetchAnalyticalEvents = async (
+  aggregator: Aggregator,
+  aggregationSpecs: AggregationSpecs,
+  fetchOptions: FetchOptions,
+) => {
+  const aggregationsByCode = getAggregationsByCode(aggregationSpecs);
+  const analytics = await fetchAnalytics(aggregator, aggregationsByCode, fetchOptions);
+  const events = analyticsToAnalyticalEvents(analytics);
+
+  const allElements = Object.keys(aggregationsByCode);
+  const checkEventIncludesAllElements = (event: AnalyticalEvent) =>
+    allElements.every(member => Object.keys(event.dataValues).includes(member));
+
+  // Remove events that do not include all elements referenced in the specs
+  return events.filter(checkEventIncludesAllElements);
+};
+
+const buildAnalyticValues = (analyticalEvents: AnalyticalEvent[], formula: string) =>
+  analyticalEvents.map(({ organisationUnit, period, dataValues }) => ({
+    organisationUnit,
+    period,
+    value: runArithmetic(formula, dataValues),
+  }));
+
 export const buildArithmetic: Builder<ArithmeticConfig> = async (input: {
   aggregator: Aggregator;
   config: ArithmeticConfig;
   fetchOptions: FetchOptions;
-}) => {
+}): Promise<AnalyticValue[]> => {
   const { aggregator, config, fetchOptions } = input;
-  validateConfig(config);
-
   const { formula, aggregation: aggregationSpecs } = config;
-  const aggregationsByCode = getAggregationsByCode(aggregationSpecs);
-  const analytics = await fetchAnalytics(aggregator, aggregationsByCode, fetchOptions);
 
-  // TODO implement
-  return [];
+  validateConfig(config);
+  const events = await fetchAnalyticalEvents(aggregator, aggregationSpecs, fetchOptions);
```

Comment:
Now that we have an implementation of the code, I thought it would be useful to clarify how analytics are matched based on `dataElement`/`orgUnit`/`period` combinations for indicator analytics calculation. Notes: https://github.com/beyondessential/tupaia-backlog/issues/255#issuecomment-681183732

We have discussed most of that stuff already, and relevant test cases are captured in `buildArithmetic.test.ts`. Something that may be slightly new is how multiple analytics for the same `dataElement`/`orgUnit`/`period` combination are handled, see the "An element is included multiple times" section in the notes

Reply:
That's good to hear! It seems that assigning names to concepts is a prerequisite for managing a higher level of abstraction